stage/stgdump.man
.\" @(#)$RCSfile: stgdump.man,v $ $Revision: 1.1 $ $Date: 2002/05/06 09:14:42 $ CERN IT/DS/HSM Jean-Damien Durand
.\" Copyright (C) 2002 by CERN/IT/DS/HSM
.\" All rights reserved
.\"
.TH STGDUMP l "$Date: 2002/05/06 09:14:42 $"
.SH NAME
stgdump \- Ask a stager deamon to dump its in\-memory catalog and save it in flat file format - for backup purpose
.SH SYNOPSIS
.B stgdump
[
.BI \-h " stghost"
]
.BI \-\-stcp_output " stgcat_dump_filename"
.BI \-\-stpp_output " stgpath_dump_filename"
.SH DESCRIPTION
.LP
The
.B stgdump
command will dial with the stager daemon and will save in disk format the content of its catalog. Output files can be HSM filenames.
See the \fBEXAMPLES\fP section below. The flat-file version of the catalog is typically for a backup purpose. See
.B stgconvert
to know how to convert a flat-file catalog to a stager database one.
.SH OPTIONS
.RS
.B \-h
stghost
.RS
Host where is running a stager deamon. Can be overwriten by STAGE_HOST environment variable.
.RE
.B \-\-stcp_output
stgcat_dump_filename
.RS
Output file for the main catalog
.RE
.B \-\-stpp_output
stgpath_dump_filename
.RS
Output file for the link catalog
.RE

.SH EXAMPLE
.ft 3
.nf
.sp
export STAGE_HOST=stagebackup
stgdump \-h stagepublic \-\-stcp_output /castor/cern.ch/c3/backup/default/2002/05/05/stgcat

will ask for a dump of in\-memory catalog of 'stagepublic' stager and use the 'stagebackup' one to store (and later migrate) the result.
.ft
.LP
.fi

.SH NOTES
If you wish to recover a catalog from a flat-file, you have to shutdown the stager, remove (or rename) the directory containing the stager catalog, typically /usr/spool/db/stage, and then use stgconvert tool and restart stager.
.SH SEE ALSO
\fBstgconvert\fP(1), \fBstgdaemon\fP(1), \fBstageshutdown\fP(1), \fBCdbserver\fP(1), 

.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>


CHANGELOG.md
## 5.11.23.0-1 (2026-09-25)

### New Features

- [frontend] Add user configuration improvements for cta-admin-grpc (cta/CTA!1635)

## 5.11.22.0-1 (2026-09-04)

### New Features

- [frontend] Implement JWT blacklisting on the CTA Frontend (cta/CTA!1548)

- [runtime] Add additional validation logic to application config (cta/CTA!1567)


### Bug Fixes

- [scheduler] Fix issues after pg1 release rest: repack admin commands, archive...

- [Tools] Fix 'priority' column in 'cta-admin dr ls' command (cta/CTA!1546)

- [runtime] Make signals, telemetry and health handling more robust (cta/CTA!1554)

- [runtime] Add bounds checking and better error diagnostics to TOML loading (cta/CTA!1559)

## 5.11.21.0-1 (2026-08-25)

### New Features

- [scheduler] Add scheduler workflow metrics to OpenTelementry (cta/CTA!1387)


### Changes

- [Tools] Move cta-admin af ch command to tf ch (cta/CTA!1525)
- [Tools] Replace cta-objectstore-tools.conf with cta-scheduler.conf (cta/CTA!1540)
- [maintd] Change scheduler.objectstore_backend_path to scheduler.config_file (cta/CTA!1518)


### Bug Fixes

- [frontend] Fix logic bug in frontend authentication and SEGFAULT in cta-admin (cta/CTA!1558)
- [Tools] Fix cta-verify-file (cta/CTA!1545)
- [Tools] Add `--storage-class` option to `cta-send-event` (cta/CTA!1542)


### Performance Improvements

- [taped] Skip block reposition request if drive is already positioned correctly (cta/CTA!1527)

## 5.11.20.0-1

### New Features

- [Tools] Add optional storage class statistics, per tape, to cta-admin tool (cta/CTA!1410)
- [Misc] Add event_name to important CTA log messages (cta/CTA!1381)

### Changes

- [Tools] Change short option for cta-admin af ch --idfile from -I to -F (cta/CTA!1453)
- [taped] Taped no longer sets its own user, group and SYS_RAWIO capability (cta/CTA!1426)
- [Misc] Improve .service and .logrotate configuration (cta/CTA!1524)
- [Misc] Modernise RPM spec metadata and file handling (cta/CTA!1507)
- [Misc] Move systemd units to `/usr/lib/systemd/system` (cta/CTA!1497)
- [Misc] Change systemd to weak RPM dependency of CTA (cta/CTA!1509)
- [Misc] Change logrotate to weak RPM dependency of CTA (cta/CTA!1505)
- [Misc] Remove RPM changelog; package CHANGELOG.md as doc (cta/CTA!1508)
- [Misc] Remove noreplace from example configuration files (cta/CTA!1506)
- [taped] Drive process crashing produces CRIT message instead of INFO (cta/CTA!1499)
- [taped] cta-eosdf no longer sources sysconfig (cta/CTA!1455)

### Bug Fixes

- [frontend] Forbid reclaiming tapes with recently deleted files (cta/CTA!1418)
- [frontend] PREPARE gRPC request shouldn't fail without a storage class (cta/CTA!1471)
- [taped] Fix crash due to accidental pointer arithmetic in exception msg (cta/CTA!1502)
- [taped] Fix opentelemetry shutdown race condition with objectstore (cta/CTA!1500)
- [scheduler]: Fix archive unmount decision logic for postgres scheduler (cta/CTA!1450)
- [frontend] Fix frontend catalogue cache for mount rules (cta/CTA!1490)
- [CI] Ensure CTA images trust CERN Root CA (cta/CTA!1535)
- [Misc] Ensure CMake correctly installs executables as PROGRAMS (cta/CTA!1523)

### Security

- [CI] Generate SBOM for each Docker image (cta/CTA!1532)

### Other

- [CI] Add system tests to cover staging tokens on EOS-CTA (cta/CTA!1493)

## 5.11.19.0-1

### Changes

- [frontend] Split Admin/WFE frontends, add mTLS (cta/CTA!1367)
- [catalogue] Forbid adding or changing encryption key of non-empty tapes (cta/CTA!1414)

### Deprecations

- [frontend] Deprecate storage class field in EOS-CTA xattrs (cta/CTA!1424)

### Other

- [CI] Add Enstore and EnstoreLarge tape tests (cta/CTA!1412)

## 5.11.18.0-2

### Bug Fixes

- [Misc] Update required runtime version of opentelemetry-cpp to v1.26.0 (cta/CTA!1370)

## 5.11.18.0-1

### New Features

- [maintd] Add catalogue and telemetry conf files to runtime directory (cta/CTA!1303)
- [Misc] Package basic JSON logging schema (cta/CTA!1304)
- [Misc] Add source location and CTA version to log lines (cta/CTA!1348)

### Changes

- [frontend] Allow admin commands or physics workflows to be disabled on CTA frontend (cta/CTA!1336)
- [Misc] Replace reflect-cpp library with custom implementation (cta/CTA!1291)
- [Misc] Change Catalogue backend install behaviour (cta/CTA!1295)

### Bug Fixes

- [catalogue] Improve CTA resilience to catalogue DB outages during data fetching (cta/CTA!1334)
- [taped] Improve readExactBlock error message contents (cta/CTA!1311)
- [taped] Enabling reading of tapes with misformed OSM tape format (cta/CTA!1343)
- [frontend] Prevent 'cta-admin dr rm' command from crashing the cta-frontend (cta/CTA!1305)
- [frontend] Remove race conditions in NegotiationRequestHandler (cta/CTA!1310)
- [frontend] Fix broken cta-admin fr ls command (cta/CTA!1340)
- [Misc] Fix compile issues when building CTA in Release mode (cta/CTA!1207)

### Performance Improvements

- [CI] Use python xrootd persistent clients for archive instead of xrdcp --parallel (cta/CTA!1274)

### Other

- [Misc] Update opentelemetry-cpp to v1.26.0 (cta/CTA!1321)


## 5.11.17.0-1

### New Features

- [maintd] Add CLI option to use runtime directory for app metadata (cta/CTA!1266)

### Changes

- [maintd] Update maintd to use TOML configuration (cta/CTA!1246)
- [maintd] Use declarative config to initialise OpenTelemetry SDK (cta/CTA!1248)
- [maintd] Add use_unix_domain_socket configuration entry for health server (cta/CTA!1282)
- [maintd] Remove daemon_user and daemon_group settings (cta/CTA!1252)

### Bug Fixes

- [frontend] Fix crash in 'cta-admin dr ls' command when using multiple...
- [frontend] Add error message when trying to change media type of non-empty tape (cta/CTA!1253)

### Removals

- [frontend] Stop using TAPE_POOL.SUPPLY column (cta/CTA!1245)

### Performance Improvements

- [taped] Optimise DriveState queries (cta/CTA!1198)

### Other

- [Misc] Rename *.conf.example files to *.example.conf (cta/CTA!1279)

## 5.11.16.1-1

### Bug Fixes


stage/stgdump.c
/*
 * $Id: stgdump.c,v 1.5 2002/11/19 08:57:32 jdurand Exp $
 */

/*
 * Copyright (C) 2002 by CERN/IT/DS/HSM
 * All rights reserved
 */

#ifndef lint
static char sccsid[] = "@(#)$RCSfile: stgdump.c,v $ $Revision: 1.5 $ $Date: 2002/11/19 08:57:32 $ CERN IT-PDP/DM Jean-Damien Durand";
#endif /* not lint */

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#ifndef _WIN32
#include <unistd.h>
#endif
#include <string.h>
#include <errno.h>
#include "stage_api.h"
#include "osdep.h"
#include "Cgetopt.h"
#include "serrno.h"
#include "rfio_api.h"

void log_callback _PROTO((int, char *));
void qry_callback _PROTO((struct stgcat_entry *, struct stgpath_entry *));
void usage _PROTO((char *));

int fd_stcp;
int fd_stpp;
int reqid;
int istcp_output = 0;
int istpp_output = 0;

char *stcp_output = NULL;
char *stpp_output = NULL;
char *stghost_input = NULL;
char *stghost_output = NULL;
static int stcp_output_flag = 0;
static int stpp_output_flag = 0;
static int stghost_input_flag = 0;
static int stghost_output_flag = 0;

int main(argc,argv)
	int argc;
	char **argv;
{
	struct stgcat_entry stcp_dummy;
	int nstcp_output = 0;
	int nstpp_output = 0;
	int rc;
	char *func = "stgdump";
	int errflg = 0;
	int c;
	char *stghost = NULL;
	int v;

	static struct Coptions longopts[] =
	{
		{"stcp_output",        REQUIRED_ARGUMENT,  &stcp_output_flag,  1},
		{"stpp_output",        REQUIRED_ARGUMENT,  &stpp_output_flag,  1},
		{"stghost_input",      REQUIRED_ARGUMENT,  &stghost_input_flag,  1},
		{"stghost_output",     REQUIRED_ARGUMENT,  &stghost_output_flag,  1},
		{NULL,                 0,                  NULL,               0}
	};

	if (stage_setlog(&log_callback) != 0) {
		stage_errmsg(func,"startup : Cannot set log callback (%s)\n", sstrerror(serrno));
	}

	Coptind = 1;
	Copterr = 1;

	stcp_output_flag = 0;
	stpp_output_flag = 0;

	while ((c = Cgetopt_long (argc, argv, "", longopts, NULL)) != -1) {
		switch (c) {
		case 'h':
			stghost = Coptarg;
			break;
		case 0:
			/* Long option without short option correspondance */
			if (stcp_output_flag && ! stcp_output) {
				/* --stcp_output */
				stcp_output = Coptarg;
			}
			if (stpp_output_flag && ! stpp_output) {
				/* --stpp_output */
				stpp_output = Coptarg;
			}
			if (stghost_input_flag && ! stghost_input) {
				/* --stghost_input */
				stghost_input = Coptarg;
			}
			if (stghost_output_flag && ! stghost_output) {
				/* --stghost_output */
				stghost_output = Coptarg;
			}
			break;
		case '?':
			errflg++;
			break;
		default:
			errflg++;
			break;
		}
        if (errflg != 0) break;
	}

	if (errflg != 0) {
		usage(argv[0]);
		exit(EXIT_FAILURE);
	}
	if ((stcp_output == NULL) && (stpp_output == NULL)) {
		stage_errmsg(func,"arguments : --stcp_output or --stpp_output are mandatory\n");
		usage(argv[0]);
		exit(EXIT_FAILURE);
	}
	if (((stghost_input != NULL) && (stghost_output == NULL)) ||
		((stghost_input == NULL) && (stghost_output != NULL))) {
		stage_errmsg(func,"arguments : --stghost_input and --stghost_output must be set all together\n");
		usage(argv[0]);
		exit(EXIT_FAILURE);
	}
	
	if ((stcp_output != NULL) && (stpp_output != NULL)) {
		if (strcmp(stcp_output,stpp_output) == 0) {
			stage_errmsg(func,"arguments : --stcp_output and --stpp_output are equivalent\n");
			exit(EXIT_FAILURE);
		}
	}
	
	/* We will not seek etc... so we can work in V3 mode */
	v = RFIO_STREAM;
	rfiosetopt(RFIO_READOPT,&v,4); 

	if (stghost_output != NULL) {
		char myenv[1024];

		sprintf(myenv,"STAGE_HOST=%s",stghost_output);
		if (putenv(myenv) != 0) {
			stage_errmsg(func,"Cannot putenv(\"%s\"), %s\n", myenv, strerror(errno));
		}
	}

	/* Open output files */
	if (stcp_output != NULL) {
		PRE_RFIO;
		if ((fd_stcp = rfio_open64(stcp_output, O_CREAT|O_WRONLY|O_TRUNC, 0644)) < 0) {
			stage_errmsg(func,"%s rfio_open64 error : %s\n", stcp_output, rfio_serror());
			exit(EXIT_FAILURE);
		}
	}
	if (stpp_output != NULL) {
		PRE_RFIO;
		if ((fd_stpp = rfio_open64(stpp_output, O_CREAT|O_WRONLY|O_TRUNC, 0644)) < 0) {
			stage_errmsg(func,"%s rfio_open64 error : %s\n", stpp_output, rfio_serror());
			rfio_close(fd_stcp);
			exit(EXIT_FAILURE);
		}
	}

	if (stage_setcallback(&qry_callback) != 0) {
		stage_errmsg(func,"startup : Cannot set callback (%s)\n", sstrerror(serrno));
	}

	memset(&stcp_dummy, 0, sizeof(struct stgcat_entry));
	if (stcp_output != NULL) {
		if (stghost_input != NULL) {
			char myenv[1024];
			
			sprintf(myenv,"STAGE_HOST=%s",stghost_input);
			if (putenv(myenv) != 0) {
				stage_errmsg(func,"Cannot putenv(\"%s\"), %s\n", myenv, strerror(errno));
			}
		}

		if ((rc = stage_qry('t',                    /* t_or_d */
							(u_signed64) STAGE_ALL, /* Flags */
							stghost,                /* Hostname */
							1,                      /* nstcp_input */
							&stcp_dummy,            /* stcp_input */
							&nstcp_output,          /* nstcp_output */
							NULL,                   /* stcp_output */
							&nstpp_output,          /* nstpp_output */
							NULL                    /* stpp_output */
			)) < 0) {
			stage_errmsg(func,"stage_qry error : %s\n", sstrerror(serrno));
		} else {
			fprintf(stdout,"==> Got %d stcp\n", nstcp_output);
		}
	}
	if (stpp_output != NULL) {
		if (stghost_input != NULL) {
			char myenv[1024];
			
			sprintf(myenv,"STAGE_HOST=%s",stghost_input);
			if (putenv(myenv) != 0) {
				stage_errmsg(func,"Cannot putenv(\"%s\"), %s\n", myenv, strerror(errno));
			}
		}
		if ((rc = stage_qry('t',                    /* t_or_d */
							(u_signed64) STAGE_ALL|STAGE_LINKNAME, /* Flags */
							stghost,                /* Hostname */
							1,                      /* nstcp_input */
							&stcp_dummy,            /* stcp_input */
							&nstcp_output,          /* nstcp_output */
							NULL,                   /* stcp_output */
							&nstpp_output,          /* nstpp_output */
							NULL                    /* stpp_output */
			)) < 0) {
			stage_errmsg(func,"stage_qry error : %s\n", sstrerror(serrno));
		} else {
			fprintf(stdout,"==> Got %d stpp\n", nstpp_output);
		}
	}

	if (stghost_output != NULL) {
		char myenv[1024];

		sprintf(myenv,"STAGE_HOST=%s",stghost_output);
		if (putenv(myenv) != 0) {
			stage_errmsg(func,"Cannot putenv(\"%s\"), %s\n", myenv, strerror(errno));
		}
	}

	if (stcp_output != NULL) {
		PRE_RFIO;
		if ((rc = rfio_close(fd_stcp)) < 0) {
			stage_errmsg(func,"%s close error : %s\n", stcp_output, rfio_serror());
		}
	}
	if (stpp_output != NULL) {
		PRE_RFIO;
		if ((rc = rfio_close(fd_stpp)) < 0) {
			stage_errmsg(func,"%s close error : %s\n", stpp_output, rfio_serror());
		}
	}

	exit(rc ? EXIT_FAILURE : EXIT_SUCCESS);

}

void log_callback(level,message)
	int level;
	char *message;
{
	if (level == MSG_ERR) {
		fprintf(stderr,"%s",message);
	}
	return;
}

void qry_callback(stcp,stpp)
	struct stgcat_entry *stcp;
	struct stgpath_entry *stpp;
{
	char *func = "qry_callback";
	if ((stcp == NULL) && (stpp == NULL)) {
		stage_errmsg(func,"(stcp == NULL) && (stpp == NULL) !?\n");
		return;
	}
	if ((stcp != NULL) && (stcp_output != NULL)) {
		/* stcp record */
		PRE_RFIO;
		if (rfio_write(fd_stcp, stcp, (int) sizeof(struct stgcat_entry)) != sizeof(struct stgcat_entry)) {
			stage_errmsg(func,"write of stcp error : %s\n", rfio_serror());
			exit(EXIT_FAILURE);
		} else {
			if (++istcp_output % 1000 == 0) fprintf(stdout,"... Got %d stcp\n", istcp_output);
		}
		
	}
	if ((stpp != NULL) && (stpp_output != NULL)) {
		/* stpp record */
		PRE_RFIO;
		if (rfio_write(fd_stpp, stpp, (int) sizeof(struct stgpath_entry)) != sizeof(struct stgpath_entry)) {
			stage_errmsg(func,"write of stpp error : %s\n", rfio_serror());
			exit(EXIT_FAILURE);
		} else {
			if (++istpp_output % 1000 == 0) fprintf(stdout,"... Got %d stcp\n", istpp_output);
		}
		
	}
}

void usage(prog)
	char *prog;
{
	fprintf(stderr,"Usage: %s [--stghost_input hostname] [--stghost_output hostname] --stcp_output <filename> --stcpp_output <filename>\n", prog);
}
